jxl/src/render/mod.rs

```diff
@@ -0,0 +1,135 @@
+    //    corresponding output pixels.
+    fn process_row_chunk(
+        &mut self,
+        position: (usize, usize),
+        xsize: usize,
+        input: &[&[&[InputT]]],
+        output: &mut [&mut [&mut [OutputT]]],
+    );
+}
+
+// Does not directly modify the current image pixels, but extends the current image with
+// additional data.
+pub trait RenderPipelineExtendStage<T: ImageDataType> {
+    // uses_channel is assumed to be always true
+    fn new_size(&self) -> (usize, usize);
+    fn original_data_origin(&self) -> (usize, usize);
+    // passed buffer must always be entirely outside of the original image.
+    fn fill_padding_row_chunk(
+        &mut self,
+        new_position: (usize, usize),
+        xsize: usize,
+        row: &mut [&mut [T]],
+    );
+}
+
+pub trait RenderPipelineBuilder: Sized {
+    type RenderPipeline: RenderPipeline;
+
+    fn new(num_channels: usize, size: (usize, usize), group_size: usize) -> Self;
+
+    fn add_input_stage<T: ImageDataType, Stage: RenderPipelineInputStage<T>>(
+        self,
+        stage: Stage,
+    ) -> Result<Self>;
+
+    fn add_inplace_stage<T: ImageDataType, Stage: RenderPipelineInPlaceStage<T>>(
+        self,
+        stage: Stage,
+    ) -> Result<Self>;
+
+    fn add_inout_stage<
+        InT: ImageDataType,
+        OutT: ImageDataType,
+        Stage: RenderPipelineInOutStage<InT, OutT>,
+    >(
+        self,
+        stage: Stage,
+    ) -> Result<Self>;
+
+    fn add_extend_stage<T: ImageDataType, Stage: RenderPipelineExtendStage<T>>(
+        self,
+        stage: Stage,
+    ) -> Result<Self>;
+
+    fn build(self) -> Self::RenderPipeline;
+}
+
+#[allow(dead_code)]
+pub struct GroupFillInfo<F> {
+    group_id: usize,
+    num_filled_passes: usize,
+    fill_fn: F,
+}
+
+fn fake_fill_fn(_: &mut [ImageRectMut<u8>]) {
+    panic!("can only use fill_input_same_type if the inputs are of the same type");
+}
+
+pub trait RenderPipeline {
+    type Builder: RenderPipelineBuilder<RenderPipeline = Self>;
+
+    // Feeds input into the pipeline. Takes as input a vector specifying, for each group that will
+    // be filled in, how many passes are filled and how to fill in each channel.
+    fn fill_input_same_type<T: ImageDataType, F>(&mut self, group_fill_info: Vec<GroupFillInfo<F>>)
+    where
+        F: FnOnce(&mut [ImageRectMut<T>]) + Send,
+    {
+        self.fill_input_two_types(
+            group_fill_info
+                .into_iter()
+                .map(|x| GroupFillInfo {
+                    fill_fn: (x.fill_fn, fake_fill_fn),
+                    group_id: x.group_id,
+                    num_filled_passes: x.num_filled_passes,
+                })
+                .collect(),
+        );
+    }
+
+    // Same as fill_input_same_type, but the inputs might have different types.
+    // Which type is used for which channel is determined by the stages in the render pipeline.
```

Comment:
I think the comments could be actual doc comments (`/// ...`) so that docs are picked up by rustdoc and IDEs.